src/tools/wasm-fuzz-types.cpp

```diff
@@ -189,6 +197,292 @@ void Fuzzer::checkLUBs(const std::vector<HeapType>& types) {
   }
 }
 
+void Fuzzer::checkCanonicalization() {
+  // Check that structural canonicalization is working correctly by building the
+  // types again, choosing randomly between equivalent possible children for
+  // each definition from both the new and old sets of built types.
+  if (getTypeSystem() == TypeSystem::Nominal) {
+    // No canonicalization to check.
+    return;
+  }
+
+  TypeBuilder builder(types.size());
+
+  // Helper for creating new definitions of existing types, randomly choosing
+  // between canonical and temporary components.
+  struct Copier {
+    Random& rand;
+    const std::vector<HeapType>& types;
+    TypeBuilder& builder;
+
+    // For each type, the indices in `types` at which it appears.
+    std::unordered_map<HeapType, std::vector<Index>> typeIndices;
```

Comment:
Maybe `SmallVector<1>`?

Reply:
Repeats aren't super uncommon because the fuzzer goes slightly out of its way to make them more likely, so I think that would document the wrong intent.

Reply:
Ah, I see, sgtm then.